Development practices/folder_structure.md

```diff
@@ -0,0 +1,286 @@
+│   ├── exporter.rb
+│   ├── o_auth_creator.rb
+│   ├── retrieve_all_data
+│   │   ├── base.rb
+│   │   ├── favorites.rb
+│   │   ├── friends_list.rb
+│   │   ├── notifications.rb
+│   │   ├── personal_information.rb
+│   │   └── reviews.rb
+│   ├── retrieve_all_data.rb
+│   └── settings.rb
+├── vat_checker.rb
+├── version_converter.rb
+└── x_server
+    ├── add_detour_distance.rb
+    ├── add_driving_distance.rb
+    ├── base.rb
+    ├── build_polyline.rb
+    ├── corridor_search.rb
+    ├── get_detour_distance.rb
+    ├── get_driving_distance.rb
+    ├── get_geo_keywords.rb
+    ├── get_location_attributes.rb
+    ├── get_route.rb
+    └── leave_reachable.rb
+```
+
+</details>
+
+At some point this folder became a dumpall folder without rhyme or reason.
+There is also an issue of boundries: there are none. Each class is used from anywhere in the application and that can
+cause headaches.
+
+# Desired structure
+
+Here is a folder structure we should strive for
+
+```
+app
+├── controllers  # Rails 
+├── frontend     # webpacker
+├── domains      # custom
+├── models       # Rails
+├── views        # Rails
+└── web          # custom
+```
+
+Lets first talk about Rails specific folders (`controllers`, `views` and `models`).
+
+`views` folder needs to exist because it is hardcoded in the Rails itself (https://github.com/rails/rails/blob/main/railties/lib/rails/engine.rb#L610)
+
+`controllers` is there because of the way rails finds controller names (it appends `"Controller"` to the resource name) (https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/http/request.rb#L88)
+
+`models` is a folder where we put all our ActiveRecord models. It is technically not necessary but we will talk about it in [Further Improvements](#further-improvements) chapter.
+
+
+# `domains` vs `web` folder
+`web` is for classes used in controller actions: forms, policies, serializers, queries and similar.
+`domains` is for business logic.
+
+## `web` folder
+`web` folder structure should follow your routes. Having these routes:
+
+```ruby
+namespace: :dasboard do
+  resources :projects
+end
+namespace :api do
+  namespace :v1 do
+    resources :projects
+  end
+end
+```
+
+it would have this folder structure:
+
+```
+web
+├── dashboard
+│   ├── projects
+│   │   ├── presenter.rb
+│   │   ├── decorator.rb
+│   │   ├── form.rb
+│   │   └── policy.rb
+├── api
+│   ├── v1
+│   │   ├── v1
+│   │   │   ├── projects
+│   │   │   │   ├── policy.rb
+│   │   │   │   ├── query.rb
+│   │   │   │   └── serialzier.rb
+```
+
+## `domains` folder
+`domains` folder should be structured into domains. 
+
+The main idea is to create cohesive groups of domain objects that are involved in the same business use-case(s). The same way we try to build classes that are loosely coupled but highly cohesive we should build our domains as well.
+Here is an example of domains folder:
+
+```
+domains
+├── exports
+│   ├── formatters
+│   ├── csvs
+│   └── spreadsheets
+├── external
```

Comment:
For all external services, I would have an `externals` folder outside the `domains` folder. 
In that case:

- it forces me to have separation of concerns, so I do not mix internal business logic with external staff
- it forces me to have an internal API of how to use that service especially in the case service is used on multiple places or different parts of the business domain
- we have a list of all external services in one place

Reply:
But it cant be namespaced in `External` module this way :(

Reply:
I do see why it is important to be namespaced in `External` module